File: 数组/29_顺时针打印矩阵.js
// 输入一个矩阵，按照从外向里以顺时针的顺序依次打印出每一个数字。

// 示例 1：
// 输入：matrix = [[1,2,3],[4,5,6],[7,8,9]]
// 输出：[1,2,3,6,9,8,7,4,5]

// 示例 2：
// 输入：matrix = [[1,2,3,4],[5,6,7,8],[9,10,11,12]]
// 输出：[1,2,3,4,8,12,11,10,9,5,6,7]

// 类似于蛇形数组
var spiralOrder = function(matrix) {
    var arrow = [[0,1],[1,0],[0,-1],[-1,0]]
    var n = matrix.length 
    var n1 = matrix[0].length
    var i = 0,j = 0
    var res = []
    var flag = 0
    while(res.length < n*n1){
        res.push(matrix[i][j])
        matrix[i][j] = null
        var nextI = i + arrow[flag%4][0]
        var nextJ = j + arrow[flag%4][1]
        if(nextI > n-1 || nextJ > n1-1 || nextI < 0 || nextJ < 0 || matrix[nextI][nextJ] === null){
            flag ++ 
        }
        i += arrow[flag%4][0]
        j += arrow[flag%4][1]
    }
    return res
};

console.log(spiralOrder([[1,2,3,4],[5,6,7,8],[9,10,11,12]]))

File: 数组/29_顺时针打印矩阵.py
# class Solution:
#     def spiralOrder(self, matrix):
#         res = []
#         while(matrix): 
#             res += matrix.pop(0)

#             matrix = list(zip(*matrix))[::-1]
#         return res

# so = Solution()
# res = so.spiralOrder([[1,2,3],[4,5,6],[7,8,9]])
# print(res)

def spiralOrder(matrix):
    if not matrix or not matrix[0]:
        return list()
    
    rows, columns = len(matrix), len(matrix[0])
    order = list()
    left, right, top, bottom = 0, columns - 1, 0, rows - 1
    while left <= right and top <= bottom:
        for column in range(left, right + 1):
            order.append(matrix[top][column])
        for row in range(top + 1, bottom + 1):
            order.append(matrix[row][right])
        print(left,right,top,bottom)
        if left < right and top < bottom:
            for column in range(right - 1, left, -1):
                order.append(matrix[bottom][column])
            for row in range(bottom, top, -1):
                order.append(matrix[row][left])
        left, right, top, bottom = left + 1, right - 1, top + 1, bottom - 1
    return order

res = spiralOrder([[1,2,3],[4,5,6],[7,8,9],[10,11,12]])